### PR TITLE
Use ginkgo labels for marking tests

### PR DESCRIPTION
**Pull request by divyaKh**
**What this PR does / why we need it**:  This PR focuses on using ginkgo label decorators to mark if certain tests belong to a specific test lane or not. The tests under consideration are:

1. `GPU` tests in test-lane `gpu`
2. `MediatedDevices` tests in test lane `vgpu`
3. `Migration` tests (which need at least 2 nodes and a CPU manager) in test lane `sig-compute-migrations`
4. `Non-root` tests in if non-root test lane

The PR adds new labels to the group of tests written in the test files mentioned below:

1. Label- `needs-gpu` in file `kubevirt/tests/vmi_gpu_test.go`
2. Label- `needs-mdev-gpu` in file `kubevirt/tests/mdev_configuration_allocation_test.go`. 
3. Label- `needs-two-nodes-with-cpumanager` in file `kubevirt/tests/migration_test.go`.
5. Label- `needs-non-root` in file `kubevirt/tests/nonroot_test.go`.

This eliminates the dependency on the use of random keywords matchers.
--label-filter is then used with gingko params in automation/test.sh, to indicate all test lanes that require these tests.

**Which issue(s) this PR fixes** :

**Special notes for your reviewer**:

**Release note**:

```release-note

```

### CODE REVIEW

File: automation/test.sh [hunk 1]
```diff
@@ -372,27 +372,33 @@ if [[ -z ${KUBEVIRT_E2E_FOCUS} && -z ${KUBEVIRT_E2E_SKIP} ]]; then
     export KUBEVIRT_E2E_FOCUS="\\[sig-storage\\]|\\[storage-req\\]"
     export KUBEVIRT_E2E_SKIP="Migration"
   elif [[ $TARGET =~ vgpu.* ]]; then
-    export KUBEVIRT_E2E_FOCUS=MediatedDevices
+    export KUBEVIRT_E2E_FOCUS="MediatedDevices"
```
Comment (dhiller) [inline automation/test.sh:375]: Not sure whether the same label selector would need to get applied here. @xpivarc might know more than me.

```suggestion
    export KUBEVIRT_E2E_FOCUS="MediatedDevices"
    export ginko_params="$ginko_params --label-filter=\"needs-gpu\""
```

File: tests/vmi_gpu_test.go [hunk 2]
```diff
@@ -54,7 +54,7 @@ func checkGPUDevice(vmi *v1.VirtualMachineInstance, gpuName string) {
 	Expect(err).ToNot(HaveOccurred(), "GPU device %q was not found in the VMI %s within the given timeout", gpuName, vmi.Name)
 }
 
-var _ = Describe("[Serial][sig-compute]GPU", func() {
+var _ = Describe("[Serial][sig-compute]GPU", Label("needs-gpu"), func() {
```
Comment (dhiller) [inline tests/vmi_gpu_test.go:57]: @xpivarc we might need to add the same label to the `MediatedDevices` tests, no?

File: automation/test.sh (hunk 1 above)
Comment (divyaKh) [reply automation/test.sh:375]: @xpivarc Your thoughts?

File: automation/test.sh [hunk 3]
```diff
@@ -372,27 +372,33 @@ if [[ -z ${KUBEVIRT_E2E_FOCUS} && -z ${KUBEVIRT_E2E_SKIP} ]]; then
...
-    export KUBEVIRT_E2E_SKIP="GPU|MediatedDevices|Migration"
+    export KUBEVIRT_E2E_SKIP="MediatedDevices|Migration"
+    export ginko_params="$ginko_params --label-filter=\"!needs-gpu\""
   elif [[ $TARGET =~ sig-monitoring ]]; then
       export KUBEVIRT_E2E_FOCUS="\\[sig-monitoring\\]"
   elif [[ $TARGET =~ sig-operator ]]; then
     export KUBEVIRT_E2E_FOCUS="\\[sig-operator\\]"
   elif [[ $TARGET =~ sriov.* ]]; then
-    export KUBEVIRT_E2E_FOCUS=SRIOV
+    export KUBEVIRT_E2E_FOCUS="SRIOV"
```
Comment (EdDev) [inline automation/test.sh:391]: Adding the braces in this commit makes it harder to review only the parts that actually are relevant to dropping the GPU focus. Could you please extract such changes to a different commit?

File: automation/test.sh [hunk 4]
```diff
@@ -372,27 +372,33 @@ if [[ -z ${KUBEVIRT_E2E_FOCUS} && -z ${KUBEVIRT_E2E_SKIP} ]]; then
...
   elif [[ $TARGET =~ gpu.* ]]; then
-    export KUBEVIRT_E2E_FOCUS=GPU
+    export KUBEVIRT_E2E_FOCUS=""
```
Comment (EdDev) [inline automation/test.sh:393]: What is the reason for setting this env variable now?

File: automation/test.sh (hunk 1 above)
Comment (EdDev) [reply automation/test.sh:375]: > Not sure whether the same label selector would need to get applied here. @xpivarc might know more than me.

I do not understand the question.
This is a different marking, appearing in `mdev_configuration_allocation_test.go`.

File: docs/test-lanes.md [hunk 5]
```diff
@@ -0,0 +1,6 @@
+#TEST LANES LABELS
```
Comment (EdDev) [inline docs/test-lanes.md:1]: There is a README file under the `tests` folder, I think it is more appropriate to add this information there.

But I am not sure the information here is relevant for running the tests using `make functest` locally.
There are two "API" for running the e2e tests, one from the CI and one locally. If you want to provide documentation about it, there is a need to clarify which one you use and how one should use it all the way through (i.e. what to call and with what parameters).

**Review by EdDev — COMMENTED** [review]
I like the usage of the labels and the steps to deprecate the old markers.

I would also add a print that warns about this deprecation so other users will be notified that the deprecation is under way.
You will also need to notify the kubevirt community (though the mailing list I guess) of this move.

File: automation/test.sh (hunk 4 above)
Comment (divyaKh) [reply automation/test.sh:393]: @EdDev I had an understanding that since it is being used in _/kubevirt/hack/functests.sh_ , so it required to initialize it (that is why I kept it as an empty string). But I think that initializing **$KUBEVIRT_E2E_FOCUS** is not compulsory, I would remove this env variable as indicated by you above. Thanks for bring that up!

File: automation/test.sh (hunk 3 above)
Comment (divyaKh) [reply automation/test.sh:391]: @EdDev By extracting the changes, you mean should I modify that commit? - like in here (https://stackoverflow.com/questions/1186535/how-to-modify-a-specified-commit)

**Comment by divyaKh** [comment]
> deprecate the old markers

Added the print statement. Thanks for suggesting that.

File: docs/test-lanes.md (hunk 5 above)
Comment (divyaKh) [reply docs/test-lanes.md]: @EdDev I have updated the `tests/README.md` and deleted the 'docs/test-lanes.md`.

File: tests/README.md [hunk 6]
```diff
@@ -119,3 +119,24 @@ docker run \
...
+
+This adds new labels to the group of tests written in the test files mentioned below:
+
+1. Label- `needs-gpu` in file `kubevirt/tests/vmi_gpu_test.go`
+2. Label- `needs-mdev-gpu` in file `kubevirt/tests/mdev_configuration_allocation_test.go`. 
+
+This eliminates the dependency on the use of random keywords matchers.
+--label-filter is then used with gingko params in automation/test.sh, to indicate all test lanes that require these tests.
+
+**Note**: For these tests to pass, the TARGET must have atleast one node on the cluster that contains an Nvidia GPU. For mediated devices tests, requirement is of an Nvidia GPU + mediated device.
```
Comment (fossedihelm) [inline tests/README.md:138]: nit: missing space "at<space>least"

Comment (divyaKh) [reply tests/README.md]: @fossedihelm  resolved. Thanks

**Comment by fossedihelm** [comment]
@divyaKh I found the error that causes `sig-compute` failures, before you removed the `needs-mdev` label.
The error was caused by a wrong escape of the label-filter inserting them inside https://github.com/kubevirt/kubevirt/blob/6bf6ca9c19cc9be11b3bbeb5ad3b8bc1c35f2bc3/hack/functests.sh#L66
This because conditional filters is a mix of key char in linux: `!` for history expansion, `&&` as operator... 

That `$@` result command to be executed like:
```
_out/tests/ginkgo -timeout=3h -r -slow-spec-threshold=60s --nodes=3 '--skip=\[Serial\]|Migration|\[verify-nonroot\]|QUARANTINE' '--focus=\[sig-compute\]' --no-color --seed=42 '--label-filter=!needs-mdev-gpu' '&&' '!needs-gpu' _out/tests/tests.test
```
So only `!needs-mdev-gpu` is evaluated.
After many tries, the only solution I found is to use a different specific variable for the label filter, here is a commit example:
https://github.com/fossedihelm/kubevirt/commit/77d5c0af21151bddfe41a45506dbc0269221a97f

and with this the result will be:
```
_out/tests/ginkgo -timeout=3h -r -slow-spec-threshold=60s --nodes=3 '--skip=\[Serial\]|Migration|\[verify-nonroot\]|QUARANTINE' '--focus=\[sig-compute\]' --no-color --seed=42 '--label-filter=!needs-gpu && !needs-mdev-gpu' _out/tests/tests.test
```
Hope this helps

**Comment by divyaKh** [comment]
> @divyaKh I found the error that causes `sig-compute` failures, before you removed the `needs-mdev` label. The error was caused by a wrong escape of the label-filter inserting them inside
> 
> https://github.com/kubevirt/kubevirt/blob/6bf6ca9c19cc9be11b3bbeb5ad3b8bc1c35f2bc3/hack/functests.sh#L66
> 
> 
> This because conditional filters is a mix of key char in linux: `!` for history expansion, `&&` as operator...
> That `$@` result command to be executed like:
> 
> ```
> _out/tests/ginkgo -timeout=3h -r -slow-spec-threshold=60s --nodes=3 '--skip=\[Serial\]|Migration|\[verify-nonroot\]|QUARANTINE' '--focus=\[sig-compute\]' --no-color --seed=42 '--label-filter=!needs-mdev-gpu' '&&' '!needs-gpu' _out/tests/tests.test
> ```
> 
> So only `!needs-mdev-gpu` is evaluated. After many tries, the only solution I found is to use a different specific variable for the label filter, here is a commit example: [fossedihelm@77d5c0a](https://github.com/fossedihelm/kubevirt/commit/77d5c0af21151bddfe41a45506dbc0269221a97f)
> 
> and with this the result will be:
> 
> ```
> _out/tests/ginkgo -timeout=3h -r -slow-spec-threshold=60s --nodes=3 '--skip=\[Serial\]|Migration|\[verify-nonroot\]|QUARANTINE' '--focus=\[sig-compute\]' --no-color --seed=42 '--label-filter=!needs-gpu && !needs-mdev-gpu' _out/tests/tests.test
> ```
> 
> Hope this helps


Thank you @fossedihelm ! As discussed, the changes have been incorporated.

**Comment by divyaKh** [comment]
@fossedihelm @jean-edouard  Can you review this PR?

**Comment by xpivarc** [comment]
/close

Thank you for the work.
This was now merged as part of https://github.com/kubevirt/kubevirt/pull/8881
